### PR TITLE
fix: fix files validation on file uploader, add unique ID

### PR DESCRIPTION
**Pull request by JKMarkowski**
#### Please provide a link to the associated issue.
part of https://github.com/SAP/fundamental-ngx/issues/3085
#### Please provide a brief summary of this pull request.
There is added unique ID for every single file uploader, if ID is not set, to avoid repeated inputs.
There is added same validation for file DND directive and FileUploader Component. The API for file uploader component is same, the only breaking change is in file dnd
Breaking Changes - `file-uploader-dragndrop.directive.ts` removed `invalidFileDrop` event, `fileChanged` event now throws all of dropped files wrapped in proper interface 
```
{
    validFiles?: File[];
    invalidFiles?: File[];
}
```

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

### CODE REVIEW

File: libs/core/src/lib/file-uploader/file-uploader.component.ts [hunk 1]
```diff
@@ -179,15 +180,24 @@ export class FileUploaderComponent implements ControlValueAccessor {
...
 
-        this.validFiles = event.filter(file => checkSize(file.size));
-        this.invalidFiles = event.filter(file => !checkSize(file.size));
+        let allowedExtensions = null;
+        if (this.accept) {
+            allowedExtensions = this.accept.toLocaleLowerCase().replace(/[\s.]/g, '').split(',')
+        }
 
+        this.validFiles = event.filter(file => checkSize(file.size) && this._checkExtension(file, allowedExtensions));
+        this.invalidFiles = event.filter(file => !checkSize(file.size) || !this._checkExtension(file, allowedExtensions));
```
Comment (DeepakSap14) [inline libs/core/src/lib/file-uploader/file-uploader.component.ts:200]: should we not change style, if invalid file is getting uploaded.
like we do in other form component, we set state = 'error' or 'warning'.

Comment (JKMarkowski) [reply libs/core/src/lib/file-uploader/file-uploader.component.ts:200]: For sure state shouldn't be changed under the hood

Comment (JKMarkowski) [reply libs/core/src/lib/file-uploader/file-uploader.component.ts:200]: I tried to add some example with states, but `readonly` state of input has it's own styling and it overwrites the semantic states(warning, success etc.). We should contact designer how it should look like in this case

File: libs/core/src/lib/file-uploader/file-uploader.component.ts [hunk 2]
```diff
@@ -245,4 +255,11 @@ export class FileUploaderComponent implements ControlValueAccessor {
         this.inputRef.nativeElement.value = '';
         this.onChange([]);
     }
+
+    /** @hidden */
+    private _checkExtension(file: File, allowedExtensions?: string[]): boolean {
```
Comment (salarenko) [inline libs/core/src/lib/file-uploader/file-uploader.component.ts:260]: In what situation is `allowedExtensions` optional?

**Review by salarenko — COMMENTED** [review]
LGTM, one question:

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
For the invalid upload example, when I drag a png (should be valid), I get both a valid alert and an invalid alert.

Also for min/max file size, if i drag a file that is <1kb, i get the invalid alert twice

File: libs/core/src/lib/file-uploader/file-uploader.component.ts (hunk 2 above)
Comment (stefanoScalzo) [reply libs/core/src/lib/file-uploader/file-uploader.component.ts:260]: Im assuming the case where all extensions are allowed

File: libs/core/src/lib/file-uploader/file-uploader.component.ts [hunk 3]
```diff
@@ -179,15 +180,24 @@ export class FileUploaderComponent implements ControlValueAccessor {
         }
         const maxSize = this.parseFileSize(this.maxFileSize);
         const minSize = this.parseFileSize(this.minFileSize);
+
         const checkSize = (fileSize: number): boolean => {
-            if (this.maxFileSize && fileSize > maxSize) { return false; }
-            if (this.minFileSize && fileSize < minSize) { return false; }
+            if (this.maxFileSize && fileSize > maxSize) {
```
Comment (stefanoScalzo) [inline libs/core/src/lib/file-uploader/file-uploader.component.ts:185]: if(this.maxFileSize && fileSize > maxSize || this.minFileSize && fileSize < minSize) {
return false;
}

Comment (JKMarkowski) [reply libs/core/src/lib/file-uploader/file-uploader.component.ts]: It has been moved to service, I sticked to 2 ifs, because it's easier to read

**Comment by JKMarkowski** [comment]
@mikerodonnell89 This componet has been refactored, I moved everything to separated service, now file DND directive and fileUploader component got same validation. It works well atm.

File: apps/docs/src/app/core/component-docs/file-uploader/examples/file-uploader-max-example/file-uploader-max-example.component.ts [hunk 4]
```diff
@@ -9,14 +9,12 @@ export class FileUploaderMaxExampleComponent {
     files: File[];
     invalidFiles: File[];
 
-    handleFileSection(): void {
-
-        alert('Files selected successfully!!!');
-
+    handleFileSection(files: File[]): void {
+        alert(files.length + ' Files selected successfully!!!');
```
Comment (InnaAtanasova) [inline apps/docs/src/app/core/component-docs/file-uploader/examples/file-uploader-max-example/file-uploader-max-example.component.ts:13]: I don't think this should be hardcoded. What if there's i18n and you need to display this in another language?

File: apps/docs/src/app/core/component-docs/file-uploader/examples/file-uploader-min-and-max-size-example/file-uploader-min-and-max-size-example.component.ts [hunk 5]
```diff
@@ -9,13 +9,13 @@ export class FileUploaderMinAndMaxSizeExampleComponent {
...
+    handleFileSection(files: File[]): void {
+        alert(files.length + ' Files selected successfully!!!');
     }
 
-    handleInvalidFiles($event): void {
-        alert(' Invalid file selection ');
-        this.invalidFiles = $event;
+    handleInvalidFiles(files: File[]): void {
+        alert(files.length + ' Invalid files selected ');
+        this.invalidFiles = files;
```
Comment (InnaAtanasova) [inline apps/docs/src/app/core/component-docs/file-uploader/examples/file-uploader-min-and-max-size-example/file-uploader-min-and-max-size-example.component.ts:18]: same as my other comment about i18n. At least the application developer should be able to customize how they want to handle these 2 functions.

File: libs/core/src/lib/file-uploader/file-uploader.component.ts (hunk 2 above)
Comment (JKMarkowski) [reply libs/core/src/lib/file-uploader/file-uploader.component.ts]: It shouldn't be there. Anyway it has been moved to another service

**Review by salarenko — CHANGES_REQUESTED** [review]
1. This example:
- Is inconsistent (button text and actual `accept` parameter)
- Does not accept `.png` files
image
Example `.png` file: [image-2.zip](https://github.com/SAP/fundamental-ngx/files/5245268/image-2.zip)

2. I've already asked about it in some other PR, but this is still very confusing for me, why do we have:
- `FileUploaderDragndropDirective.disabled`
- `FileUploaderDragndropDirective.dragndrop`
if both are responsible for disabling drag and drop functionality?

3. `FileUploaderDragndropDirective.disabled` still emits on `dragEntered` and `dragLeave` events.

4. Does it make sense to perform any file validation in `FileUploaderDragndropDirective`?
It's good to split the responsibility between specific components. This directive should rather be responsible for emitting drag/drop related events and any other component specific logic should be handled outside.

5. What is this 😄? How can the user set the files using Template/Reactive forms if this is not handled?
image

6. We have separate examples for:
- Max and Min file size
- Max file size
- Min file size
I think we can reduce them to Max Min example.

File: apps/docs/src/app/core/component-docs/file-uploader/examples/file-uploader-min-and-max-size-example/file-uploader-min-and-max-size-example.component.ts (hunk 5 above)
Comment (InnaAtanasova) [reply apps/docs/src/app/core/component-docs/file-uploader/examples/file-uploader-min-and-max-size-example/file-uploader-min-and-max-size-example.component.ts:18]: You are not missing anything :) I know these are in the examples, but don't you think the application developers would like to handle the success/failure cases in a similar way? Like provide an API for this?

**Comment by salarenko** [comment]
@JKMarkowski 
Maybe lets move all of the issues/concerns with file-uploader file to new issue and tackle it as a separate task.
I think first point from my list "1. This example..." is in scope of this PR and should be fixed here.

**Comment by JKMarkowski** [comment]
@salarenko 
@InnaAtanasova 
Some Minor changes in docs have been introduced, let's create another task with issues listed, I wouldn't put those as a scope of this PR

**Review by mikerodonnell89 — COMMENTED** [review]
I don't think we need 3 examples for min/max file size.  But this looks good

**Review by salarenko — APPROVED** [review]

**Review by InnaAtanasova — APPROVED** [review]

**Review by mikerodonnell89 — APPROVED** [review]

**Review by droshev — APPROVED** [review]

**Pull request merged**